File: FT/Firmware/Control frame plotting/plot_display_signals.py
## Plotting display control frame
#
# *************************************************************************
#
#  @file    plot_display_signals.py
#  @author  AW    adrian.wojcik@put.poznan.pl
#  @version 2.0
#  @date    02-Jun-2022
#  @brief   Common sinal plotting function (see: Final task)
#  @remark  Testes for Python 3.10
#           Requires: numpy, matplotlib
#
# *************************************************************************

import numpy as np
import matplotlib.pyplot as plt

def plot_display_signals(TIME, CLOCK, DATA1, DATA2, DATA3, DATA4, TS, TITLE):

    #plt.rcParams.update({'font.size': 7})
    hFig, hAxs = plt.subplots(5, 1)
    hFig.canvas.manager.set_window_title(TITLE)
    plt.subplots_adjust(hspace=0.5)

    hAxs[0].set_title(TITLE,fontweight='bold')
    hAxs[0].stairs(CLOCK, TIME, linewidth=2, color='#000000')
    hAxs[0].grid(True)
    hAxs[0].set_ylabel('Logic state')
    hAxs[0].set_xlim([TIME[1],TIME[-1]])
    hAxs[0].set_ylim([-0.1,1.1])  
    hAxs[0].legend(['CLOCK'])

    hAxs[1].stairs(DATA1, TIME, linewidth=2, color='#FF0000')
    hAxs[1].grid(True)
    hAxs[1].set_ylabel('Logic state')
    hAxs[1].set_xlim([TIME[1],TIME[-1]])
    hAxs[1].set_ylim([-0.1,1.1])  
    hAxs[1].legend(['DATA1'])
    for i in np.arange(1,8,2,dtype=int):
        hAxs[1].text(TIME[2*i-1] + TS/3, 0.4, 'A'+str(int((i+1)/2)));
        hAxs[1].text(TIME[2*i+1] + TS/3, 0.4, 'E'+str(int((i+1)/2)));

    hAxs[2].stairs(DATA2, TIME, linewidth=2, color='#77AC30')
    hAxs[2].grid(True)
    hAxs[2].set_ylabel('Logic state')
    hAxs[2].set_xlim([TIME[1],TIME[-1]])
    hAxs[2].set_ylim([-0.1,1.1])  
    hAxs[2].legend(['DATA2'])
    for i in np.arange(1,8,2,dtype=int):
        hAxs[2].text(TIME[2*i-1] + TS/3, 0.4, 'B'+str(int((i+1)/2)));
        hAxs[2].text(TIME[2*i+1] + TS/3, 0.4, 'F'+str(int((i+1)/2)));

    hAxs[3].stairs(DATA3, TIME, linewidth=2, color='#0072BD')
    hAxs[3].grid(True)
    hAxs[3].set_ylabel('Logic state')
    hAxs[3].set_xlim([TIME[1],TIME[-1]])
    hAxs[3].set_ylim([-0.1,1.1])  
    hAxs[3].legend(['DATA3'])
    for i in np.arange(1,8,2,dtype=int):
        hAxs[3].text(TIME[2*i-1] + TS/3, 0.4, 'C'+str(int((i+1)/2)));
        hAxs[3].text(TIME[2*i+1] + TS/3, 0.4, 'G'+str(int((i+1)/2)));

    hAxs[4].stairs(DATA4, TIME, linewidth=2, color='#EDB120')
    hAxs[4].grid(True)
    hAxs[4].set_ylabel('Logic state')
    hAxs[4].set_xlabel('Time [ms]')
    hAxs[4].set_xlim([TIME[1],TIME[-1]])
    hAxs[4].set_ylim([-0.1,1.1])  
    hAxs[4].legend(['DATA4'])
    for i in np.arange(1,8,2,dtype=int):
        hAxs[4].text(TIME[2*i-1] + TS/3, 0.4, 'D'+str(int((i+1)/2)));
        hAxs[4].text(TIME[2*i+1] + TS/3, 0.4, 'H'+str(int((i+1)/2)));
    
    plt.show()

File: FT/Firmware/Control frame plotting/octave/plot_display_signals.m
%% Plotting display control signals
%
% *************************************************************************
%
%  @file    plot_display_signals.m
%  @author  AW    adrian.wojcik@put.poznan.pl
%  @version 2.0
%  @date    01-Jun-2022
%  @brief   Common sinal plotting function (see: Final task)
%  @remark  Testes for Octave 7.1.0
%
% *************************************************************************
function hFig = plot_display_signals(TIME, CLOCK, DATA1, DATA2, DATA3, DATA4, TS, TITLE)

% Result plotting
hFig = figure('Name', TITLE);

LW = 2; % [pt]

subplot(5, 1, 1)
    stairs(TIME, CLOCK, 'Color', '#000000', 'LineWidth', LW)
    hold on
    grid on
    ylabel('Logic state')
    title(TITLE)
    axis([TIME(1) TIME(end) -0.1 1.1])
    legend('CLOCK')
subplot(5, 1, 2)
    stairs(TIME, DATA1, 'Color', '#FF0000', 'LineWidth', LW)
    hold on
    grid on
    ylabel('Logic state')
    axis([TIME(1) TIME(end) -0.1 1.1])
    legend('DATA1')
    for i = 1 : 2 : 7
        text(TIME(2*i-1) + TS/3, 0.5, ['A' num2str((i+1)/2)]);
        text(TIME(2*i+1) + TS/3, 0.5, ['E' num2str((i+1)/2)]);
    end
subplot(5, 1, 3)
    stairs(TIME, DATA2, 'Color', '#77AC30', 'LineWidth', LW)
    hold on
    grid on
    ylabel('Logic state')
    axis([TIME(1) TIME(end) -0.1 1.1])
    legend('DATA2')
    for i = 1 : 2 : 7
        text(TIME(2*i-1) + TS/3, 0.5, ['B' num2str((i+1)/2)]);
        text(TIME(2*i+1) + TS/3, 0.5, ['F' num2str((i+1)/2)]);
    end
subplot(5, 1, 4)
    stairs(TIME, DATA3, 'Color', '#0072BD', 'LineWidth', LW)
    hold on
    grid on
    ylabel('Logic state')
    axis([TIME(1) TIME(end) -0.1 1.1])
    legend('DATA3')
    for i = 1 : 2 : 7
        text(TIME(2*i-1) + TS/3, 0.5, ['C' num2str((i+1)/2)]);
        text(TIME(2*i+1) + TS/3, 0.5, ['G' num2str((i+1)/2)]);
    end
subplot(5, 1, 5)
    stairs(TIME, DATA4, 'Color', '#EDB120', 'LineWidth', LW)
    hold on
    grid on
    ylabel('Logic state')
    xlabel('Time [ms]')
    axis([TIME(1) TIME(end) -0.1 1.1])
    legend('DATA4')
    for i = 1 : 2 : 7
        text(TIME(2*i-1) + TS/3, 0.5, ['D' num2str((i+1)/2)]);
        text(TIME(2*i+1) + TS/3, 0.5, ['H' num2str((i+1)/2)]);
    end
